language/evm/move-to-yul/src/generator.rs

```diff
@@ -0,0 +1,770 @@
+        }
+        entry_bb
+    }
+
+    /// Compute a map from labels to block ids which those labels start
+    fn compute_label_map(
+        cfg: &StacklessControlFlowGraph,
+        code: &[Bytecode],
+    ) -> BTreeMap<Label, BlockId> {
+        let mut map = BTreeMap::new();
+        for id in cfg.blocks() {
+            if let BlockContent::Basic { lower, .. } = cfg.content(id) {
+                if let Bytecode::Label(_, l) = &code[*lower as usize] {
+                    map.insert(*l, id);
+                }
+            }
+        }
+        map
+    }
+
+    /// Generate Yul statement for a bytecode.
+    fn bytecode(
+        &mut self,
+        ctx: &Context,
+        target: &FunctionTarget,
+        label_map: &BTreeMap<Label, BlockId>,
+        bc: &Bytecode,
+        has_flow: bool,
+    ) {
+        use Bytecode::*;
+        let fun = target.func_env;
+        let print_loc = || {
+            let loc = target.get_bytecode_loc(bc.get_attr_id());
+            emitln!(
+                ctx.writer,
+                "/// @src {}:{}:{}",
+                ctx.env.file_id_to_idx(loc.file_id()),
+                loc.span().start(),
+                loc.span().end()
+            );
+        };
+        let print_nyi = || {
+            emitln!(
+                ctx.writer,
+                "// NYI: {}",
+                bc.display(target, &BTreeMap::default())
+            );
+        };
+        let get_block = |l| label_map.get(l).expect("label has corresponding block");
+        let local_name = |l: &TempIndex| ctx.make_local_name(fun, *l);
+        let mut builtin = |yul_fun: YulFunction, dest: &[TempIndex], srcs: &[TempIndex]| {
+            print_loc();
+            emitln!(
+                ctx.writer,
+                "{} := {}",
+                local_name(&dest[0]),
+                self.call_builtin_str(ctx, yul_fun, srcs.iter().map(local_name))
+            )
+        };
+        let mut builtin_typed = |yul_fun_u8: YulFunction,
+                                 yul_fun_u64: YulFunction,
+                                 yul_fun_u128: YulFunction,
+                                 dest: &[TempIndex],
+                                 srcs: &[TempIndex]| {
+            use PrimitiveType::*;
+            use Type::*;
+            match target.get_local_type(srcs[0]) {
+                Primitive(U8) => builtin(yul_fun_u8, dest, srcs),
+                Primitive(U64) => builtin(yul_fun_u64, dest, srcs),
+                Primitive(U128) => builtin(yul_fun_u128, dest, srcs),
+                _ => panic!("unexpected operand type"),
+            }
+        };
+        match bc {
+            Jump(_, l) => {
+                print_loc();
+                emitln!(ctx.writer, "$block := {}", get_block(l))
+            }
+            Branch(_, if_t, if_f, cond) => {
+                print_loc();
+                emitln!(
+                    ctx.writer,
+                    "if {} {{ $block := {} }} else {{ $block := {} }}",
```

Comment:
Unfortunately, no "else" block can be defined in Yul. Instead, "switch" is needed: https://docs.soliditylang.org/en/latest/yul.html#if

Reply:
Done